src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,519 @@
+        self.is_async = (self.hdr_flags & FLAG_ASYNC) > 0
+        self.is_related = (self.hdr_flags & FLAG_RELATED) > 0
+        self.is_signed = (self.hdr_flags & FLAG_SIGNED) > 0
+        # FIXME other flags 3.1 or too obscure
+        if self.is_async:
+            (self.async_id, self.session_id,
+             self.signature) = struct.unpack("<QQ16s", packet[32:64])
+            self.tree_id = 0x00
+        else:
+            (_reserved, self.tree_id, self.session_id,
+             self.signature) = struct.unpack("<LLQ16s", packet[32:64])
+            self.async_id = 0x00
+        log.debug("HEADER")
+        log.debug("------")
+        log.debug("protocol ID     {pid!r}", pid=protocol_id)
+        log.debug("size            {hs}", hs=hdr_size)
+        log.debug("credit charge   {cc}", cc=self.credit_charge)
+        log.debug("status          {status}", status=hdr_status)
+        log.debug("command         {cmd!r} {cmdn:02x}",
+                  cmd=COMMANDS[self.hdr_command],
+                  cmdn=self.hdr_command)
+        log.debug("credit request  {cr}", cr=self.credit_request)
+        s = ""
+        if self.is_async:
+            s += "ASYNC "
+        if self.is_signed:
+            s += "SIGNED "
+        if self.is_related:
+            s += "RELATED "
+        log.debug("flags           0x{flags:x} {s}", flags=self.hdr_flags, s=s)
+        log.debug("next command    0x{nc:x}", nc=self.next_command)
+        log.debug("message ID      0x{mid:x}", mid=self.message_id)
+        log.debug("session ID      0x{sid:x}", sid=self.session_id)
+        if self.is_async:
+            log.debug("async ID        0x{aid:x}", aid=self.async_id)
+        else:
+            log.debug("tree ID         0x{tid:x}", tid=self.tree_id)
+        log.debug("signature       {sig}",
+                  sig=binascii.hexlify(self.signature))
+        try:
+            name, req_type, resp_type = COMMANDS[self.hdr_command]
+        except IndexError:
+            log.error("unknown command 0x{cmd:x}", cmd=self.hdr_command)
+            self.error_response(STATUS_NOT_IMPLEMENTED)
+        else:
+            func = 'smb_' + name
+            try:
+                if hasattr(self, func) and req_type:
+                    req = req_type(packet[64:])
+                    resp = resp_type()
+                    getattr(self, func)(req, resp)
+                else:
+                    log.error("command '{cmd}' not implemented",
+                              cmd=COMMANDS[self.hdr_command][0])
+                    self.error_response(STATUS_NOT_IMPLEMENTED)
+            except base.SMBError as e:
+                log.error(str(e))
+                self.error_response(e.ntstatus)
+            except BaseException:
+                log.failure("in {cmd}", cmd=COMMANDS[self.hdr_command][0])
+                self.error_response(STATUS_UNSUCCESSFUL)
+        if self.is_related and self.next_command > 0:
+            self.packetReceived(packet[self.next_command:])
```

Comment:
Tail recursion is risky if these packets might get big; you can end up blowing the recursion limit. Could this be unwound to be iterable instead?